doc/crypto/api/ops/ka.rst

```diff
@@ -114,6 +122,90 @@ Key agreement algorithms
 Standalone key agreement
 ------------------------
 
+.. function:: psa_key_agreement
+
+    .. summary::
+        Perform a key agreement and return the shared secret as a derivation key.
+
+    .. param:: psa_algorithm_t alg
+        The key agreement algorithm to compute: a value of type `psa_algorithm_t` such that :code:`PSA_ALG_IS_RAW_KEY_AGREEMENT(alg)` is true.
+    .. param:: psa_key_id_t private_key
+        Identifier of the private key to use.
+        It must permit the usage `PSA_KEY_USAGE_DERIVE`.
+    .. param:: const uint8_t * peer_key
+        Public key of the peer. The peer key must be in the same format that `psa_import_key()` accepts for the public key type corresponding to the type of ``private_key``. That is, this function performs the equivalent of :code:`psa_import_key(..., peer_key, peer_key_length)`, with key attributes indicating the public key type corresponding to the type of ``private_key``. For example, for ECC keys, this means that peer_key is interpreted as a point on the curve that the private key is on. The standard formats for public keys are documented in the documentation of `psa_export_public_key()`.
+    .. param:: size_t peer_key_length
+        Size of ``peer_key`` in bytes.
+
+    .. param:: const psa_key_attributes_t * attributes
+        The attributes for the new key.
+        This function uses the attributes as follows:
+
+        *   The key type must either be `PSA_KEY_TYPE_NONE` (``0``), or `PSA_KEY_TYPE_DERIVE`. :issue:`(Is this ok, or should we allow other unstructured (or just raw?) key types to be returned? Or should we mandate exactly one of those options, because the function always returns KEY_TYPE DERIVE?)`
+        *   The key size must either be ``0``, or the size of the key agreement output, in bits. :issue:`(Is this ok, or should we mandate leaving as default zero? Or should we permit the secret to be truncated into the output key?)`
+
+            The output size, in bits, of the key agreement is :code:`8 * PSA_RAW_KEY_AGREEMENT_OUTPUT_SIZE(type, bits)`, where ``type`` is the type of ``private_key`` and ``bits`` is the bit-size of either ``private_key`` or the ``peer_key``.
```

Comment:
> Truncating the output is unnecessary, and risks making the shared secret discoverable.

I actually wonder. For example, after a key exchange on secp521r1, you get a 521-bit number which is close to uniform over the range 0..2^521-1, but when forced into 8-bit bytes, you get a 66-byte string where the first byte's top bits are 0. So if you use this value, it may well make sense to just remove the _first_ byte.

You can truncate by exporting the shared secret and fiddling with it in the application space. But the whole point of this new function is that the shared secret remains in the secure world when the crypto implementation has isolation.

I'm wary of having customized truncation because this is basically creating a Spectre gadget.

Reply:
I basically think we shouldn't do anything with the shared secret before the shared secret is derived. So we don't have the problem of the first bit being zero. The truncation shall be after the key derivation.

Reply:
I've updated the PR to only support using the entire shared secret as the output key, and requiring the attributes to specify either zero or the correct number of bits (as per `psa_import_key()` and `psa_copy_key()`).